client/client.go

```diff
@@ -1521,6 +1547,34 @@ func (c *Client) setupNode() error {
 		node.Meta["connect.proxy_concurrency"] = defaultConnectProxyConcurrency
 	}
 
+	// Merge dynamic node metadata
+	dynamicMeta, err := c.stateDB.GetNodeMeta()
+	if err != nil {
+		return fmt.Errorf("error reading dynamic node metadata: %w", err)
+	}
+	for dk, dv := range dynamicMeta {
+		if dv == nil {
+			_, ok := node.Meta[dk]
+			if ok {
+				// Unset static node metadata
+				delete(node.Meta, dk)
+			} else {
+				// Forget dynamic node metadata tombstone as there's no
+				// static value to erase.
+				delete(dynamicMeta, dk)
```

Comment:
Why throw this value away? It doesn't have an impact on the resulting merged metadata, but if we discard the tombstone then a subsequent config update (from restart) can cause the field to reappear without any trace of having discarded it previously. That seems like it could be confusing. (And the operator can always submit a new `nomad node config set` to wipe out any unwanted tombstones, but then that's under their control.)

Reply:
> And the operator can always submit a new `nomad node config set` to wipe out any unwanted tombstones, but then that's under their control.

(`nomad node meta apply` now). They can't, at least right now. Since `unset` and `empty` are both valid values for node metadata keys, we don't have a distinct delete/purge/forget/use-static option yet... This pruning seemed simpler than explaining the difference between `unset` and a hypothetical `delete` but it seems more like I'm making *implicit* complexity instead of *explicit* complexity.

I think implicit pruning is fairly safe because it only does something unexpected in the following situation:

1. Node has `foo="i'll be back"` set in HCL
2. Node has `foo=null` from the API
3. *Operator removes `foo` from HCL entirely*
4. Agent restarts, pruning occurs
5. Operator adds `foo="helo"` back into HCL
6. Agent restarts and `foo="helo"` appears

I just have trouble imagining a case in which this happens and the result is particularly surprising. If anything it seems most likely to occur when considerable time passes between 4 & 5, and the end result is what the operator intends (`foo` to come back)... but idk... I can also see how this could be magical and confusing.

Reply:
Fair enough. This solution effectively garbage-collects unused keys too. :+1: